appvia/kore
# Introduce service credentials + operator

## What

This PR introduces a new API object: ServiceCredentials.

The ServiceCredentials object should be created by users, if they want to provision service credentials for a specific cluster + namespace and have it as a Secret object in their cluster, so they can mount it in their applications.

Example manifest:
```
apiVersion: services.kore.appvia.io/v1
kind: ServiceCredentials
metadata:
  name: foo
  namespace: a-team
spec:
  cluster:
    group: clusters.compute.kore.appvia.io
    kind: Cluster
    name: a-team-gke-development
    namespace: a-team
    version: v1
  clusterNamespace: dev
  configuration:
    bar: baz
  kind: dummy
  service:
    group: services.kore.appvia.io
    kind: Service
    name: dummy
    namespace: a-team
    version: v1
```

ServiceCredentials can also have configurations, e.g. to create different credentials for a read-only vs. an admin user. The configuration is validated with a JSON schema. (The dummy service credentials require a `bar` configuration value)

There is also an operator which will either request the provider to provision credentials and will save it as `Secret` object in the target cluster's namespace, or on delete it will delete the `Secret` object and will ask the provider to revoke the credentials.

## Additional changes

 * Add optional Credentials allocation to services (I didn't add this in the first PR, but I realised we will need this, e.g. to authenticate with a cloud provider or with a service broker)
 * Always check whether a team exists in the CLI (we get wrong errors if the team doesn't exist, and we don't check that in many team related commands)

## Testing

1. Enable the `services` feature gate (see https://github.com/appvia/kore/pull/673)
1. Create a GKE cluster with a namespace, e.g. `dev`
1. Create a dummy service
    ```
    kore create service dummy --plan dummy --team a-team
    ```
1. Create service credentials
    ```
    kore create servicecredentials foo --team a-team --cluster a-team-gke-development --service dummy --namespace dev --param bar=baz
    ```
1. Refresh the kube configs
    ```
    kore kubeconfig --team a-team
    ```
1. Check if the secret exists in the target cluster and namespace. The `data.superSecret` value should be `foo-secret` (base64 encoded).
    ```
    kubectl get secret foo -n dev -o yaml
    ```
1. Delete the secret:
    ```
    kore delete servicecredentials foo --team a-team
    ```
1. Make sure it's gone in the Kubernetes cluster
    ```
    kubectl get secret foo -n
    ```
